File: domain.guice/src/main/java/org/openmedstack/domain/guice/EventStoreModule.java
package org.openmedstack.domain.guice;

import com.google.inject.AbstractModule;
import com.google.inject.Injector;
import org.openmedstack.IProvideTenant;
import org.openmedstack.domain.Memento;
import org.openmedstack.domain.Saga;
import org.openmedstack.eventstore.*;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.Charset;
import java.time.Instant;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public class EventStoreModule extends AbstractModule {

    @Override
    protected void configure() {
        bind(IConstructAggregates.class).to(ContainerAggregateFactory.class).asEagerSingleton();
        bind(IConstructSagas.class).to(ContainerSagaFactory.class).asEagerSingleton();
        bind(IStoreEvents.class).to(HttpEventStore.class).asEagerSingleton();
        bind(IAccessSnapshots.class).to(HttpSnapshotStore.class).asEagerSingleton();
        bind(Repository.class).toConstructor((Constructor<DelegateAggregateRepository>)DelegateAggregateRepository.class.getConstructors()[0]).asEagerSingleton();
        bind(SagaRepository.class).toConstructor((Constructor<DelegateSagaRepository>)DelegateSagaRepository.class.getConstructors()[0]).asEagerSingleton();
        bind(IProvideTenant.class).toInstance(new IProvideTenant() {
            @Override
            public String getTenantName() {
                return "test";
            }
        });
        bind(IDetectConflicts.class).to(ConflictDetector.class);
    }
}

File: domain.guice/src/main/java/org/openmedstack/domain/guice/EventStoreModule.kt
package org.openmedstack.domain.guice

import com.google.inject.AbstractModule
import org.openmedstack.IProvideTenant
import org.openmedstack.eventstore.*
import java.lang.reflect.Constructor

class EventStoreModule : AbstractModule() {
    override fun configure() {
        bind(IConstructAggregates::class.java).to(ContainerAggregateFactory::class.java)
        bind(IConstructSagas::class.java).to(ContainerSagaFactory::class.java)
        bind(IStoreEvents::class.java).to(HttpEventStore::class.java)
        bind(IAccessSnapshots::class.java).to(HttpSnapshotStore::class.java)
        bind(Repository::class.java).toConstructor(
            DelegateAggregateRepository::class.java.getConstructor(
                IStoreEvents::class.java,
                IAccessSnapshots::class.java,
                IConstructAggregates::class.java
            )
        )
        bind(SagaRepository::class.java).toConstructor(
            DelegateSagaRepository::class.java.getConstructor(
                IProvideTenant::class.java,
                IStoreEvents::class.java,
                IConstructSagas::class.java
            )
        )
        bind(IProvideTenant::class.java).toInstance(object : IProvideTenant {
            override val tenantName: String
                get() = "test"
        })
        bind(IDetectConflicts::class.java).to(ConflictDetector::class.java)
    }
}
